File: Tests/java/TestRegions.java
/*
 [The "BSD license"]
 Copyright (c) 2009 Terence Parr
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package org.stringtemplate.v4.test;

import org.junit.Test;
import org.stringtemplate.v4.ST;
import org.stringtemplate.v4.STErrorListener;
import org.stringtemplate.v4.STGroup;
import org.stringtemplate.v4.STGroupFile;
import org.stringtemplate.v4.misc.ErrorBuffer;

import static org.junit.Assert.assertEquals;

public class TestRegions extends BaseTest {
    @Test public void testEmbeddedRegion() throws Exception {
        String dir = getRandomDir();
        String groupFile =
            "a() ::= <<\n" +
            "[<@r>bar<@end>]\n" +
            ">>\n";
        writeFile(dir, "group.stg", groupFile);
        STGroup group = new STGroupFile(dir+"/group.stg");
        ST st = group.getInstanceOf("a");
        String expected = "[bar]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testRegion() throws Exception {
        String dir = getRandomDir();
        String groupFile =
            "a() ::= <<\n" +
            "[<@r()>]\n" +
            ">>\n";
        writeFile(dir, "group.stg", groupFile);
        STGroup group = new STGroupFile(dir+"/group.stg");
        ST st = group.getInstanceOf("a");
        String expected = "[]";
        String result = st.render();
        assertEquals(expected, result);
    }

	@Test public void testDefineRegionInSubgroup() throws Exception {
		String dir = getRandomDir();
		String g1 = "a() ::= <<[<@r()>]>>\n";
		writeFile(dir, "g1.stg", g1);
		String g2 = "@a.r() ::= <<foo>>\n";
		writeFile(dir, "g2.stg", g2);

		STGroup group1 = new STGroupFile(dir+"/g1.stg");
		STGroup group2 = new STGroupFile(dir+"/g2.stg");
		group2.importTemplates(group1); // define r in g2
		ST st = group2.getInstanceOf("a");
		String expected = "[foo]";
		String result = st.render();
		assertEquals(expected, result);
	}

    @Test public void testDefineRegionInSubgroupThatRefsSuper() throws Exception {
        String dir = getRandomDir();
        String g1 = "a() ::= <<[<@r>foo<@end>]>>\n";
        writeFile(dir, "g1.stg", g1);
        String g2 = "@a.r() ::= <<(<@super.r()>)>>\n";
        writeFile(dir, "g2.stg", g2);

        STGroup group1 = new STGroupFile(dir+"/g1.stg");
        STGroup group2 = new STGroupFile(dir+"/g2.stg");
        group2.importTemplates(group1); // define r in g2
        ST st = group2.getInstanceOf("a");
        String expected = "[(foo)]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testDefineRegionInSubgroup2() throws Exception {
        String dir = getRandomDir();
        String g1 = "a() ::= <<[<@r()>]>>\n";
        writeFile(dir, "g1.stg", g1);
        String g2 = "@a.r() ::= <<foo>>>\n";
        writeFile(dir, "g2.stg", g2);

        STGroup group1 = new STGroupFile(dir+"/g1.stg");
        STGroup group2 = new STGroupFile(dir+"/g2.stg");
        group1.importTemplates(group2); // opposite of previous; g1 imports g2
        ST st = group1.getInstanceOf("a");
        String expected = "[]"; // @a.r implicitly defined in g1; can't see g2's
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testDefineRegionInSameGroup() throws Exception {
        String dir = getRandomDir();
        String g = "a() ::= <<[<@r()>]>>\n"+
                   "@a.r() ::= <<foo>>\n";
        writeFile(dir, "g.stg", g);

        STGroup group = new STGroupFile(dir+"/g.stg");
        ST st = group.getInstanceOf("a");
        String expected = "[foo]";
        String result = st.render();
        assertEquals(expected, result);
    }

	@Test public void testAnonymousTemplateInRegion() throws Exception {
		String dir = getRandomDir();
		String g = "a() ::= <<[<@r()>]>>\n" +
				   "@a.r() ::= <<\n" +
				   "<[\"foo\"]:{x|<x>}>\n" +
				   ">>\n";
		writeFile(dir, "g.stg", g);

		STGroup group = new STGroupFile(dir+"/g.stg");
		ST st = group.getInstanceOf("a");
		String expected = "[foo]";
		String result = st.render();
		assertEquals(expected, result);
	}

    @Test public void testCantDefineEmbeddedRegionAgain() throws Exception {
        String dir = getRandomDir();
        String g = "a() ::= <<[<@r>foo<@end>]>>\n"+
                   "@a.r() ::= <<bar>>\n"; // error; dup
        writeFile(dir, "g.stg", g);

        STGroupFile group = new STGroupFile(dir+"/g.stg");
        ErrorBuffer errors = new ErrorBuffer();
        group.setListener(errors);
        group.load();
        String expected = "g.stg 2:3: region a.r is embedded and thus already implicitly defined"+newline;
        String result = errors.toString();
        assertEquals(expected, result);
    }

	@Test public void testIndentBeforeRegionIsIgnored() throws Exception {
		String dir = getRandomDir();
		String g = "a() ::= <<[\n" +
				   "  <@r>\n" +
				   "  foo\n" +
				   "  <@end>\n" +
				   "]>>\n";
		writeFile(dir, "g.stg", g);

		STGroupFile group = new STGroupFile(dir+"/g.stg");
		ST st = group.getInstanceOf("a");
		String expected = "[\n" +
						  "  foo\n" +
						  "]";
		String result = st.render();
		assertEquals(expected, result);
	}

	@Test public void testRegionOverrideStripsNewlines() throws Exception {
		String dir = getRandomDir();
		String g =
				"a() ::= \"X<@r()>Y\"" +
				"@a.r() ::= <<\n" +
				"foo\n" +
				">>\n";
		writeFile(dir, "g.stg", g);
		STGroupFile group = new STGroupFile(dir+"/g.stg");

		String sub =
				"@a.r() ::= \"A<@super.r()>B\"" +newline;
		writeFile(dir, "sub.stg", sub);
		STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
		subGroup.importTemplates(group);

		ST st = subGroup.getInstanceOf("a");
		String result = st.render();
		String expecting = "XAfooBY";
		assertEquals(expecting, result);
	}

    //

    @Test public void testRegionOverrideRefSuperRegion() throws Exception {
        String dir = getRandomDir();
        String g =
                "a() ::= \"X<@r()>Y\"" +
                "@a.r() ::= \"foo\"" +newline;
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");

        String sub =
                "@a.r() ::= \"A<@super.r()>B\"" +newline;
        writeFile(dir, "sub.stg", sub);
        STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
        subGroup.importTemplates(group);

        ST st = subGroup.getInstanceOf("a");
        String result = st.render();
        String expecting = "XAfooBY";
        assertEquals(expecting, result);
    }

    @Test public void testRegionOverrideRefSuperRegion3Levels() throws Exception {
        String dir = getRandomDir();
        // Bug: This was causing infinite recursion:
        // getInstanceOf(super::a)
        // getInstanceOf(sub::a)
        // getInstanceOf(subsub::a)
        // getInstanceOf(subsub::region__a__r)
        // getInstanceOf(subsub::super.region__a__r)
        // getInstanceOf(subsub::super.region__a__r)
        // getInstanceOf(subsub::super.region__a__r)
        // ...
        // Somehow, the ref to super in subsub is not moving up the chain
        // to the @super.r(); oh, i introduced a bug when i put setGroup
        // into STG.getInstanceOf()!

        String g =
                "a() ::= \"X<@r()>Y\"" +
                "@a.r() ::= \"foo\"" +newline;
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");

        String sub =
                "@a.r() ::= \"<@super.r()>2\"" +newline;
        writeFile(dir, "sub.stg", sub);
        STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
        subGroup.importTemplates(group);

        String subsub =
                "@a.r() ::= \"<@super.r()>3\"" +newline;
        writeFile(dir, "subsub.stg", subsub);
        STGroupFile subSubGroup = new STGroupFile(dir+"/subsub.stg");
        subSubGroup.importTemplates(subGroup);

        ST st = subSubGroup.getInstanceOf("a");

        String result = st.render();
        String expecting = "Xfoo23Y";
        assertEquals(expecting, result);
    }

    @Test public void testRegionOverrideRefSuperImplicitRegion() throws Exception {
        String dir = getRandomDir();
        String g =
                "a() ::= \"X<@r>foo<@end>Y\""+newline;
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");

        String sub =
                "@a.r() ::= \"A<@super.r()>\"" +newline;
        writeFile(dir, "sub.stg", sub);
        STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
        subGroup.importTemplates(group);

        ST st = subGroup.getInstanceOf("a");
        String result = st.render();
        String expecting = "XAfooY";
        assertEquals(expecting, result);
    }

    @Test public void testUnknownRegionDefError() throws Exception {
        String dir = getRandomDir();
        String g =
                "a() ::= <<\n" +
                "X<@r()>Y" +
                ">>\n"+
                "@a.q() ::= \"foo\"" +newline;
        STErrorListener errors = new ErrorBuffer();
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");
		group.setListener(errors);
        ST st = group.getInstanceOf("a");
        st.render();
        String result = errors.toString();
        String expecting = "g.stg 3:3: template a doesn't have a region called q"+newline;
        assertEquals(expecting, result);
    }

    @Test public void testSuperRegionRefMissingOk() throws Exception {
        String dir = getRandomDir();
        String g =
            "a() ::= \"X<@r()>Y\"" +
            "@a.r() ::= \"foo\"" +newline;
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");

        String sub =
            "@a.r() ::= \"A<@super.q()>B\"" +newline; // allow this; trap at runtime
        STErrorListener errors = new ErrorBuffer();
        group.setListener(errors);
        writeFile(dir, "sub.stg", sub);
        STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
        subGroup.importTemplates(group);

        ST st = subGroup.getInstanceOf("a");
        String result = st.render();
        String expecting = "XABY";
        assertEquals(expecting, result);
    }
}

File: Tests/objc/TestRegions.m
#import "TestRegions.h"

@implementation TestRegions

- (void) test01EmbeddedRegion
{
    NSString *dir = [self getRandomDir];
    NSString *groupFile = @"a() ::= <<\n[<@r>bar<@end>]\n>>\n";
    [self writeFile:dir fileName:@"group.stg" content:groupFile];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/group.stg", dir]] autorelease];
    ST *st = [group getInstanceOf:@"a"];
    NSString *expected = @"[bar]";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
}

- (void) test02Region
{
    NSString *dir = [self getRandomDir];
    NSString *groupFile = @"a() ::= <<\n[<@r()>]\n>>\n";
    [self writeFile:dir fileName:@"group.stg" content:groupFile];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/group.stg", dir]] autorelease];
    ST *st = [group getInstanceOf:@"a"];
    NSString *expected = @"[]";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
}

- (void) test03DefineRegionInSubgroup
{
    NSString *dir = [self getRandomDir];
    NSString *g1 = @"a() ::= <<[<@r()>]>>\n";
    NSString *g2 = @"@a.r() ::= <<foo>>\n";
    [self writeFile:dir fileName:@"g1.stg" content:g1];
    [self writeFile:dir fileName:@"g2.stg" content:g2];
    STGroup *group1 = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g1.stg", dir]] autorelease];
    STGroup *group2 = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g2.stg", dir]] autorelease];
    [group2 importTemplates:group1];
    ST *st = [group2 getInstanceOf:@"a"];
    NSString *expected = @"[foo]";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
}

- (void) test04DefineRegionInSubgroupThatRefsSuper
{
    NSString *dir = [self getRandomDir];
    NSString *g1 = @"a() ::= <<[<@r>foo<@end>]>>\n";
    NSString *g2 = @"@a.r() ::= <<(<@super.r()>)>>\n";
    [self writeFile:dir fileName:@"g1.stg" content:g1];
    [self writeFile:dir fileName:@"g2.stg" content:g2];
    STGroup *group1 = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g1.stg", dir]] autorelease];
    STGroup *group2 = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g2.stg", dir]] autorelease];
    [group2 importTemplates:group1];
    ST *st = [group2 getInstanceOf:@"a"];
    NSString *expected = @"[(foo)]";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
}

- (void) test05DefineRegionInSubgroup2
{
    NSString *dir = [self getRandomDir];
    NSString *g1 = @"a() ::= <<[<@r()>]>>\n";
    [self writeFile:dir fileName:@"g1.stg" content:g1];
    NSString *g2 = @"@a.r() ::= <<foo>>>\n";
    [self writeFile:dir fileName:@"g2.stg" content:g2];
    STGroup *group1 = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g1.stg", dir]] autorelease];
    STGroup *group2 = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g2.stg", dir]] autorelease];
    [group1 importTemplates:group2];
    ST *st = [group1 getInstanceOf:@"a"];
    NSString *expected = @"[]";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
}

- (void) test06DefineRegionInSameGroup
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= <<[<@r()>]>>\n@a.r() ::= <<foo>>\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    ST *st = [group getInstanceOf:@"a"];
    NSString *expected = @"[foo]";
    [st.impl dump];
    NSString *result = [st render];
    [self assertEquals:expected result:result];
}

- (void) test07AnonymousTemplateInRegion
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= <<[<@r()>]>>\n@a.r() ::= <<\n<[\"foo\"]:{x|<x>}>\n>>\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    ST *st = [group getInstanceOf:@"a"];
    NSString *expected = @"[foo]";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
}

- (void) test08CantDefineEmbeddedRegionAgain
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= <<[<@r>foo<@end>]>>\n@a.r() ::= <<bar>>\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
    [group setListener:errors];
    [group load];
    NSString *expected = @"g.stg 2:3: region a.r is embedded and thus already implicitly defined\n";
    NSString *result = [errors description];
    [self assertEquals:expected result:result];
}

- (void) test09CantDefineEmbeddedRegionAgainInTemplate
{
    NSString * dir = [self getRandomDir];
    NSString * g = @"a() ::= <<\n[\n<@r>foo<@end>\n<@r()>]\n>>\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile * group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    ErrorBuffer * errors = [[ErrorBuffer newErrorBuffer] autorelease];
    [group setListener:errors];
    [group load];
    NSString * expected = @"g.stg 3:2: redefinition of region a.r\n";
    NSString * result = [errors description];
    [self assertEquals:expected result:result];
}

- (void) test10MissingRegionName
{
    NSString * dir = [self getRandomDir];
    NSString * g = @"@t.() ::= \"\"\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile * group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    ErrorBuffer * errors = [[ErrorBuffer newErrorBuffer] autorelease];
    [group setListener:errors];
    [group load];
    NSString * expected = @"g.stg 1:3: missing ID at '('\n";
    NSString * result = [errors description];
    [self assertEquals:expected result:result];
}

- (void) test11IndentBeforeRegionIsIgnored
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= <<[\n  <@r>\n  foo\n  <@end>\n]>>\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    ST *st = [group getInstanceOf:@"a"];
    NSString *expected = @"[\n  foo\n]";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
}

- (void) test12RegionOverrideStripsNewlines
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= \"X<@r()>Y\"@a.r() ::= <<\nfoo\n>>\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    NSString *sub = @"@a.r() ::= \"A<@super.r()>B\"\n";
    [self writeFile:dir fileName:@"sub.stg" content:sub];
    STGroupFile *subGroup = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/sub.stg", dir]] autorelease];
    [subGroup importTemplates:group];
    ST *st = [subGroup getInstanceOf:@"a"];
    NSString *result = [st render];
    NSString *expecting = @"XAfooBY";
    [self assertEquals:expecting result:result];
}

- (void) test13RegionOverrideRefSuperRegion
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= \"X<@r()>Y\"@a.r() ::= \"foo\"\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    NSString *sub = @"@a.r() ::= \"A<@super.r()>B\"\n";
    [self writeFile:dir fileName:@"sub.stg" content:sub];
    STGroupFile *subGroup = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/sub.stg", dir]] autorelease];
    [subGroup importTemplates:group];
    ST *st = [subGroup getInstanceOf:@"a"];
    NSString *result = [st render];
    NSString *expecting = @"XAfooBY";
    [self assertEquals:expecting result:result];
}

- (void) test14RegionOverrideRefSuperRegion3Levels
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= \"X<@r()>Y\"@a.r() ::= \"foo\"\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    NSString *sub = @"@a.r() ::= \"<@super.r()>2\"\n";
    [self writeFile:dir fileName:@"sub.stg" content:sub];
    STGroupFile *subGroup = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/sub.stg", dir]] autorelease];
    [subGroup importTemplates:group];
    NSString *subsub = @"@a.r() ::= \"<@super.r()>3\"\n";
    [self writeFile:dir fileName:@"subsub.stg" content:subsub];
    STGroupFile *subSubGroup = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/subsub.stg", dir]] autorelease];
    [subSubGroup importTemplates:subGroup];
    ST *st = [subSubGroup getInstanceOf:@"a"];
    NSString *result = [st render];
    NSString *expecting = @"Xfoo23Y";
    [self assertEquals:expecting result:result];
}

- (void) test15RegionOverrideRefSuperImplicitRegion
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= \"X<@r>foo<@end>Y\"\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    NSString *sub = @"@a.r() ::= \"A<@super.r()>\"\n";
    [self writeFile:dir fileName:@"sub.stg" content:sub];
    STGroupFile *subGroup = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/sub.stg", dir]] autorelease];
    [subGroup importTemplates:group];
    ST *st = [subGroup getInstanceOf:@"a"];
    NSString *result = [st render];
    NSString *expecting = @"XAfooY";
    [self assertEquals:expecting result:result];
}

- (void) test16UnknownRegionDefError
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= <<\nX<@r()>Y>>\n@a.q() ::= \"foo\"\n";
    id<STErrorListener> errors = [ErrorBuffer newErrorBuffer];
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    [group setListener:errors];
    ST *st = [group getInstanceOf:@"a"];
    [st render];
    NSString *result = [errors description];
    NSString *expecting = @"g.stg 3:3: template a doesn't have a region called q\n";
    [self assertEquals:expecting result:result];
}

- (void) test17SuperRegionRefMissingOk
{
    NSString *dir = [self getRandomDir];
    NSString *g = @"a() ::= \"X<@r()>Y\"@a.r() ::= \"foo\"\n";
    [self writeFile:dir fileName:@"g.stg" content:g];
    STGroupFile *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/g.stg", dir]] autorelease];
    NSString *sub = @"@a.r() ::= \"A<@super.q()>B\"\n";
    id<STErrorListener> errors = [ErrorBuffer newErrorBuffer];
    [group setListener:errors];
    [self writeFile:dir fileName:@"sub.stg" content:sub];
    STGroupFile *subGroup = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/sub.stg", dir]] autorelease];
    [subGroup importTemplates:group];
    ST *st = [subGroup getInstanceOf:@"a"];
    NSString *result = [st render];
    NSString *expecting = @"XABY";
    [self assertEquals:expecting result:result];
}

- (void) test18EmbeddedRegionOnOneLine
{
    NSString * dir = [self getRandomDir];
    NSString * groupFile = @"a() ::= <<\n[\n  <@r>bar<@end>\n]\n>>\n";
    [self writeFile:dir fileName:@"group.stg" content:groupFile];
    STGroup * group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/group.stg", dir]] autorelease];
    ST * st = [group getInstanceOf:@"a"];
    [st.impl dump];
    NSString * expected = @"[\n  bar\n]";
    NSString * result = [st render];
    [self assertEquals:expected result:result];
}

- (void) test19EmbeddedRegionTagsOnSeparateLines
{
    NSString * dir = [self getRandomDir];
    NSString * groupFile = @"a() ::= <<\n[\n  <@r>\n  bar\n  <@end>\n]\n>>\n";
    [self writeFile:dir fileName:@"group.stg" content:groupFile];
    STGroup * group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/group.stg", dir]] autorelease];
    ST * st = [group getInstanceOf:@"a"];
    NSString * expected = @"[\n  bar\n]";
    NSString * result = [st render];
    [self assertEquals:expected result:result];
}

@end
